### DISCUSSION TITLE
Question on comparing ```Variable``` with integers

### DISCUSSION THREAD

**Discussion by zliucd** [Consensys/gnark#1262] [General]
Hi, I'm new to gnark, and I have a question on ```api.Cmp()```

According to comments of ```api.Cmp()```, its return value is a ```Variable``` (which is not a literal integer,  e.g, 0, 1, -1).  But how can we check the return value in ```if-else``` statement.  Below is my sample code, where I would like to check if Variable ```d``` is 0 in if-else, but it doesn't work.   It's also infeasible to convert ```Variable``` to integers (see https://github.com/Consensys/gnark/issues/1050)      

In a nutshell, how to compare ```Variable``` with integers?  Thanks! 

```
var d frontend.Variable
d = 0

api.AssertIsEqual(d, 0) // pass

if api.Cmp(d, 0) == 0 { // not print.   The right-side operand should be a Variable?
println("equal")
}

if api.Cmp(d, 0) == 1 { // not print
println("greater")
}

if api.Cmp(d, 0) == -1 { // not print
println("lesser")
}

api.Println(api.Cmp(d, 0)) // print '0'
```

Comments of ```api.Cmp()```
```
	// Cmp returns:
	//  * 1 if i1>i2,
	//  * 0 if i1=i2,
	//  * -1 if i1<i2.
	//
	// If the absolute difference between the variables i1 and i2 is known, then
	// it is more efficient to use the bounded methdods in package
	// [github.com/consensys/gnark/std/math/bits].
	Cmp(i1, i2 Variable) Variable
```

**Comment by ivokub** [comment]
The `frontend.Variable` returned is essentially any variable depending on the inputs and it does not take a value in the circuit definition. So it is not possible to define the circuit depending on the value. Instead, you should perform your operations in a way which takes into account all possible values for the variable.

For this, we have `api.Select` and `api.IsZero` methods which can be used for the parallel execution:
```go
type circuit struct {
    X, Y frontend.Variable
    Input frontend.Variable
    Ret frontend.Variable
}

func (c *circuit) Define(api frontend.API) error {
    isEqual := api.IsZero(c.X, c.Y)
    cmpRes := api.Cmp(c.X, c.Y)
    isLess := api.IsZero(api.Add(cmpRes, 1))
    isGreater := api.IsZero(api.Sub(cmpRes, 1))
    // here now only one of isEqual, isLess or IsGreater is 1
    res := api.Select(isEqual, 111, c.Input)
    res := api.Select(isLess, 222, res)
    res := api.Select(isGreater, 333, res)
    api.AssertIsEqual(res, 1)
}
```

**Reply by zliucd** [reply]
Thanks a lot. It it possible to execute function calls based on result of ```api.Select()```?  It can be a very useful feature.  Essentially, ```A``` and ```B``` are functions  of ```api.Select(result, A, B)```.  

It might be possibly done as function address can be determined at compile time, but passing params may be an issue.

**Reply by ivokub** [reply]
Doing function calls based on the select result cannot be done. F or Groth16 proving backend we encode all the circuit instructions into some equations in the form:
```
\sum a_i x_i \sum b_i x_i = \sum c_i x_i
```
where `a, b, c` depend on the circuit and `x` is the full witness. So there is no trivial way to encode functions. If you need to be able to prove general-purpose programs then I rather recommend using zkVMs a la SP1 or Risc0. But there is an overhead there in the prover and verifier cost.
